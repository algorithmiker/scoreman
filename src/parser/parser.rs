use tracing::{debug, span, trace, Level};

use super::{
    string_name,
    tab_element::{self, tab_element3, TabElement},
};
use crate::{backend::errors::backend_error::BackendError, parser::tab_element::TabElementError};
use std::ops::RangeInclusive;

pub fn line_is_valid(line: &str) -> bool {
    let line = line.trim();
    let mut chars = line.chars();
    let first_is_alphanumeric = chars.next().map(|x| x.is_alphanumeric()).unwrap_or(false);
    let second_is_measure_sep = chars.next().map(|x| x == '|').unwrap_or(false);
    let last_is_measure_end = line.ends_with('|');
    let ret = first_is_alphanumeric && second_is_measure_sep && last_is_measure_end;
    trace!(line, verdict = ret, "line_is_valid");
    ret
}

#[derive(Debug)]
pub struct Measure {
    pub data_range: RangeInclusive<u32>,
}
impl Measure {
    pub fn from(range: RangeInclusive<u32>) -> Self {
        Self { data_range: range }
    }
}

#[derive(Debug, Default)]
pub struct ParseResult {
    /// This is not a [Result] because we want to preserve the partial parse state, eg. for fixup or recovery
    pub error: Option<BackendError>,
    pub tick_stream: Vec<TabElement>,
    pub measures: Vec<Measure>,
    pub base_notes: Vec<char>,
    /// The line on which the n-th section begins and the index of the first tick in that section.
    /// This provides enough information to restore from where we have read an individual tick.
    pub offsets: Vec<(u32, u32)>,
}

impl ParseResult {
    pub fn new() -> Self {
        Self::default()
    }
    pub fn dump_tracks(&self) -> String {
        let stream_len = self.tick_stream.len();
        let tick_cnt = stream_len / 6;
        debug_assert_eq!(stream_len % 6, 0);
        let mut bufs = vec![String::new(); 6];
        for tick in 0..tick_cnt {
            let max_width =
                (0..6).map(|x| self.tick_stream[tick * 6 + x].repr_len()).max().unwrap() as usize;
            for (s, buf) in bufs.iter_mut().enumerate() {
                use tab_element::TabElement::*;
                let to_padded = |c: char| format!("{1:<0$}", max_width, c);
                match self.tick_stream[tick * 6 + s] {
                    Fret(x) => buf.push_str(&format!("{x:<0$}", max_width)),
                    Rest => buf.push_str(&to_padded('-')),
                    DeadNote => buf.push_str(&to_padded('x')),
                    Slide => buf.push_str(&to_padded('/')),
                    Bend => buf.push_str(&to_padded('b')),
                    HammerOn => buf.push_str(&to_padded('h')),
                    Pull => buf.push_str(&to_padded('p')),
                    Release => buf.push_str(&to_padded('r')),
                    Vibrato => buf.push_str(&to_padded('~')),
                }
            }
        }
        bufs.iter_mut().for_each(|x| x.push('\n'));
        bufs.concat()
    }
}

pub fn parse<T: AsRef<str>>(lines: &[T]) -> ParseResult {
    let mut r = ParseResult::new();
    let mut part_first_line = 0;
    'outer: loop {
        // find a part
        loop {
            if part_first_line + 5 >= lines.len() {
                break 'outer;
            }
            if line_is_valid(lines[part_first_line].as_ref())
                && line_is_valid(lines[part_first_line + 5].as_ref())
            {
                break;
            }
            part_first_line += 1
        }
        let range = part_first_line..=part_first_line + 5;
        let _part = span!(Level::DEBUG, "parsing part", ?range);
        let _part = _part.enter();
        r.offsets.push((part_first_line as u32, r.tick_stream.len() as u32));
        let mut part: Vec<&str> = lines[range].iter().map(|s| s.as_ref().trim()).collect(); // TODO: check if this is slow

        // The current tick in THIS PART
        let mut tick = 0;
        // parse prelude and last char
        for (line_idx, line) in part.iter_mut().enumerate() {
            let Ok((rem, string_name)) = string_name()(line) else {
                r.error = Some(BackendError::invalid_string_name(part_first_line + line_idx));
                return r;
            };
            r.base_notes.push(string_name);
            let Ok((rem, _)) = super::char('|')(rem) else {
                r.error = Some(BackendError::invalid_string_name(part_first_line + line_idx));
                return r;
            };
            *line = rem;
            if !line.ends_with('|') {
                r.error = Some(BackendError::no_closing_barline(part_first_line + line_idx));
                return r;
            };
            *line = &line[0..(line.len() - 1)];
        }

        let mut tick_cnt_est = part[0].len();
        while tick < tick_cnt_est {
            let s = span!(Level::TRACE, "parsing tick", tick);
            let _s = s.enter();
            let (mut is_multichar, mut is_multi_on) = (false, [false; 6]);
            for s in 0..6 {
                trace!(part = part[s], "remaining on string {s}:");
                if s == 0 && part[s].starts_with("|") {
                    trace!("encountered measure separator");
                    let measure_start =
                        r.measures.last().map(|x| x.data_range.end() + 1).unwrap_or(0);
                    r.measures.push(Measure::from(
                        measure_start..=r.tick_stream.len().wrapping_sub(1) as u32,
                    ));
                    part.iter_mut().for_each(|string| *string = &string[1..]); // TODO: maybe debugassert here that it is indeed a measure separator
                    tick_cnt_est -= 1;
                    trace!(part = part[s], "remaining on string {s}: after fixup");
                }

                let len_before = part[s].len();
                let (res, te) = match tab_element3(part[s]) {
                    Ok(x) => x,
                    Err((_, err)) => {
                        let (line, char) =
                            source_location_while_parsing(&r, part_first_line as u32, s as u32);
                        if let Some(TabElementError::FretTooLarge) = err {
                            r.error = Some(BackendError::large_fret(line, char));
                        } else {
                            let invalid_src = part[s].chars().next();
                            let err = BackendError::invalid_character(line, char, invalid_src);
                            r.error = Some(err);
                        }
                        return r;
                    }
                };

                let tab_element_len = len_before - res.len();
                is_multichar |= tab_element_len > 1;
                is_multi_on[s] = tab_element_len > 1;
                part[s] = res;
                r.tick_stream.push(te);
            }
            if is_multichar {
                let _ms = span!(Level::DEBUG, "marked as multichar, running fixup", tick);
                let _ms = _ms.enter();
                tick_cnt_est -= 1;
                for s in 0..6 {
                    if is_multi_on[s] {
                        trace!("multi on {s}, skipping");
                        continue;
                    };
                    let elem_idx = r.tick_stream.len() - (6 - s);
                    let elem = &r.tick_stream[elem_idx];
                    let _s = span!(Level::DEBUG, "fixing up string", string = s, ?elem);
                    let _s = _s.enter();
                    if let TabElement::Rest = elem {
                        let _s2 = span!(
                            Level::TRACE,
                            "this is a rest so we try to parse the next element"
                        );
                        let _s2 = _s2.enter();
                        let len_before = part[s].len();
                        let next = tab_element3(part[s]).unwrap(); // TODO: the unwrap here is ICE,
                                                                   // should error instead
                        if len_before - next.0.len() > 1 {
                            let (m_line, m_char) = source_location_from_stream(&r, elem_idx as u32);
                            // just for a nicer error, show another multi line too
                            let other = ((0..6).find(|x| is_multi_on[*x]).unwrap()
                                + part_first_line) as u32;
                            r.error =
                                Some(BackendError::both_slots_multichar(m_line, m_char, other));
                            return r;
                        }
                        trace!(replacement = ?next.1, "replaced a rest");
                        let len = r.tick_stream.len(); // to make the borrow checker happy about borrowing &mut and &
                        r.tick_stream[len - (6 - s)] = next.1;
                        part[s] = next.0;
                    } else {
                        let _q =
                            span!(Level::TRACE, "this is not a Rest, so we check the next element");
                        let _q = _q.enter();
                        if part[s].starts_with("-") {
                            trace!("next element is Rest so we skip it");
                            part[s] = &part[s][1..];
                        } else {
                            let (line, char) = source_location_from_stream(&r, elem_idx as u32);
                            r.error = Some(BackendError::multi_both_slots_filled(line, char));
                            return r;
                        }
                    }
                }
            }
            trace!(tick, data = r.dump_tracks(), "data after parsing tick");
            trace!(source = dump_source(&part), "source state after parsing tick");
            tick += 1;
        }

        let measure_start = r.measures.last().map(|x| x.data_range.end() + 1).unwrap_or(0);
        r.measures.push(Measure {
            data_range: measure_start..=r.tick_stream.len().wrapping_sub(1) as u32,
        });
        // finished parsing part
        trace!(part = r.dump_tracks(), "Finished part");

        part_first_line += 6;
    }
    r
}

/// A specialized, faster [source_location_from_stream]
pub fn source_location_while_parsing(
    r: &ParseResult, part_first_line: u32, line_in_part: u32,
) -> (u32, u32) {
    let actual_line = part_first_line + line_in_part;
    trace!("expecting the error to be on line {actual_line}");
    let error_tick = r.tick_stream.len() / 6;
    // we aren't accounting for measures here, so sum of all the measure lines too
    let part_start = r.offsets.last().map(|x| x.1).unwrap_or(0);
    let mut measure_lines = 0;
    for measure in r.measures.iter().rev() {
        if measure.data_range.start() < &part_start {
            break;
        }
        measure_lines += 1;
    }
    trace!("need to account for {measure_lines} measure lines");
    let mut offset_on_line = 1 + measure_lines; // e|
    let start = (part_start / 6) as usize;
    for tick in start..error_tick {
        // take the maximum extent of this tick. we cannot just add up the local tick lengths because multichars on *other strings* would throw off the parser
        // -1-2-3-
        // -11b12- <- this would think that if there is an error on the first string, the extents before are just rest-1-2-3-rest, and report an incorrect location
        let remainder = &r.tick_stream[tick * 6..];
        //trace!(?remainder);
        let tick_width = remainder.iter().take(6).map(|x| x.repr_len()).max();
        let tick_width = tick_width.unwrap_or(0);
        trace!(offset = tick_width, tick, "adding offset for tick");
        offset_on_line += tick_width;
    }
    offset_on_line += 1; // because the location refers to the offset of the tick that was not parsed
    trace!("expecting the error to be at character idx {offset_on_line}");
    (actual_line, offset_on_line)
}

pub fn source_location_from_stream(r: &ParseResult, tick_location: u32) -> (u32, u32) {
    let section = r
        .offsets
        .binary_search_by_key(&tick_location, |x| x.1)
        .unwrap_or_else(|x| x.saturating_sub(1));
    trace!("source_location_from_stream: expected to be in section {section}");
    let part_start = r.offsets[section].1;
    let idx_in_part = tick_location - part_start;
    trace!("this is the {idx_in_part}th element in the part");
    let line_in_part = idx_in_part % 6;
    let actual_line = r.offsets[section].0 + line_in_part;
    trace!("expecting the error to be on line {actual_line}");

    let error_tick = (tick_location / 6) as usize;
    // we aren't accounting for measures here, so sum of all the measure lines to
    // search for all the measures in this part, and before the needle
    let last_measure = r
        .measures
        .binary_search_by_key(&tick_location, |x| x.data_range.end() + 1)
        .unwrap_or_else(|x| x);
    debug!("last measure we need to check: {last_measure} for needle {tick_location}");
    let mut measure_lines = 0;
    trace!("{:?}", r.measures);
    trace!("part start: {part_start}");
    for (m_idx, measure) in r.measures[0..last_measure].iter().enumerate().rev() {
        if measure.data_range.start() < &part_start {
            trace!("breaking at measure {m_idx}");
            break;
        }
        measure_lines += 1;
    }
    measure_lines += 1; // for last measure; which we cannot index with 0..=last_measure if we have only 1.
    trace!("need to account for {measure_lines} measure lines");
    let mut offset_on_line = 1 + measure_lines; // e|
    let start = (part_start / 6) as usize;
    for tick in start..error_tick {
        // take the maximum extent of this tick. we cannot just add up the local tick lengths because multichars on *other strings* would throw off the parser
        // -1-2-3-
        // -11b12- <- this would think that if there is an error on the first string, the extents before are just rest-1-2-3-rest, and report an incorrect location
        let remainder = &r.tick_stream[tick * 6..];
        //traceln!(depth = 1, "remainder: {remainder:?}");
        let tick_width = remainder.iter().take(6).map(|x| x.repr_len()).max();
        let tick_width = tick_width.unwrap_or(0);
        trace!(offset = tick_width, tick, "adding offset for tick");
        offset_on_line += tick_width;
    }
    trace!("expecting the error to be at character idx {offset_on_line}");
    (actual_line, offset_on_line)
}

pub fn dump_source(input: &Vec<&str>) -> String {
    use itertools::Itertools;
    std::iter::once(&"").chain(input.iter()).join("\n")
}

#[test]
pub fn test_parse3() {
    use std::time::Instant;
    let example_score = r#"
e|--12-12|--12-12|--12-12|
B|3------|3------|3----11|
G|-6-3-3-|-6-3-3-|-6-3-11|
D|-------|-------|-----11|
A|-------|-------|-----11|
E|-----9-|-----9-|-----11|

// This is a comment!

e|--12-12|--12-12|
B|3------|3------|
G|-6-3-3-|-6-3-3-|
D|-------|-------|
A|-------|-------|
E|-----9-|-----9-|
"#;
    let lines = &example_score.lines().map(|x| x.to_string()).collect::<Vec<String>>();
    let time_parser3 = Instant::now();
    let parse3_result = parse(lines);
    println!("Parser3 took: {:?}", time_parser3.elapsed());
    insta::assert_snapshot!(parse3_result.dump_tracks());
    insta::assert_debug_snapshot!(parse3_result);
}
