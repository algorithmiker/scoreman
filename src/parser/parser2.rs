use anyhow::{bail, Context};
use std::io::BufRead;

use crate::collect_parse_error;

use super::{comment_line, partline, Score, Section};

pub fn parse2<A: BufRead>(inp: A) -> anyhow::Result<Score> {
    let mut sections = vec![];
    // Todo eventually remove Part
    let mut part_buf = vec![];
    let mut part_begin = 0;
    for (line_idx, line) in inp.lines().enumerate() {
        let line_number = line_idx + 1;
        let line = line.with_context(|| format!("Cannot read line {line_number}"))?;
        if line.trim().is_empty() {
            if !part_buf.is_empty() {
                println!("[W]: Empty line inside Part at line {line_idx}, are you sure this is intended?");
            }
            continue;
        }

        match comment_line(&line) {
            Ok((rem, comment)) => {
                if !rem.is_empty() {
                    bail!("Invalid comment syntax at line {line_number} (got remaining `{rem}`)")
                }
                if !part_buf.is_empty() {
                    println!("[W]: Comment inside Part at line {line_idx}, are you sure this is intended?");
                }
                sections.push(Section::Comment(comment.to_string()));
            }
            Err(_) => match partline(&line) {
                Ok((rem, mut line)) => {
                    if !rem.is_empty() {
                        bail!("Invalid partline at line {line_number}, got remaining content: `{rem}`");
                    }
                    // Add measure metadata
                    for measure_idx in 0..line.staffs.len() {
                        line.staffs[measure_idx].parent_line = Some(line_idx);
                        line.staffs[measure_idx].index_on_parent_line = Some(measure_idx);
                    }
                    part_buf.push(line);
                    match part_buf.len() {
                        6 => {
                            // flush part buf
                            sections.push(Section::Part {
                                part: part_buf
                                    .clone() // TODO try to elide this clone
                                    .try_into()
                                    .expect("Unreachable: more than 6 elements in part_buf"),
                                begin_line_idx: part_begin,
                                end_line_idx: line_idx,
                            });
                            part_buf.clear();
                            part_begin = 0;
                        }
                        1 => part_begin = line_idx,
                        _ => (),
                    }
                }
                // TODO maybe pass the error up here?
                Err(x) => bail!(
                    "Error at line {line_number}, not a comment but also not a valid partline\n{}",
                    collect_parse_error(x)
                ),
            },
        }
    }

    Ok(Score(sections))
}

#[test]
fn test_parse2() {
    use std::io::BufReader;
    let i1 = r#"
e|---|
B|-3-|
G|6-6|
D|---|
A|---|
E|---|

// This is a comment 

e|---|
B|3-3|
G|-6-|
D|---|
A|---|
E|---|"#;
    insta::assert_debug_snapshot!(parse2(BufReader::new(i1.as_bytes())));
    let i2 = r#"
e|---|
B|-3-|
G|6-6|
D|---|
A|-o-|
E|---|

// This is a comment 

e|---|
B|3-3|
G|-6-|
D|---|
A|---|
E|---|"#;
    insta::assert_debug_snapshot!(parse2(BufReader::new(i2.as_bytes())));
}
