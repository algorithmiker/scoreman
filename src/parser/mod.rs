pub mod parser2;
mod parser3;
#[cfg(test)]
mod parser_tests;

use crate::{digit_cnt_u8, rlen};
use parser2::BendTargets;
use std::ops::RangeInclusive;

#[derive(Debug, PartialEq)]
pub enum Section {
    Part { part: [Partline; 6] },
    Comment(String),
}

fn comment_line(s: &str) -> Result<(&str, &str), &str> {
    if s.len() < 2 || &s[0..2] != "//" {
        return Err(s);
    }
    let mut len = 0;
    for c in s.chars() {
        if c == '\n' || c == '\r' {
            break;
        }
        len += 1;
    }
    Ok((&s[len..], &s[0..len]))
}

#[derive(Debug, PartialEq, Clone)]
pub struct Partline {
    pub string_name: char,
    /// which measures originate from this partline in the measure buf of string_name
    pub measures: RangeInclusive<usize>,
}
impl Partline {
    /// Returns the measure count of this partline
    pub fn len(&self) -> usize {
        rlen(&self.measures)
    }

    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}
/// like `e|--------------4-----------|-----0--------------5-----|`
/// If called with append_to, the returned Partline will have no measures itself
/// TODO: make this know less by using callbacks
fn partline<'a>(
    s: &'a str,
    parent_line_idx: usize,
    start_source_offset: u32,
    string_buf: &mut Vec<RawTick>,
    string_measure_buf: &mut Vec<Measure>,
    string_offsets_buf: &mut Vec<u32>,
    bend_targets: &mut BendTargets,
    line_in_part: usize,
    track_measures: bool,
) -> Result<(&'a str, Partline), &'a str> {
    let (rem, string_name) = string_name()(s)?;
    let (mut rem, _) = char('|')(rem)?;
    let mut last_parsed_idx: u32 = 1;
    let mut measures = string_measure_buf.len()..=string_measure_buf.len();

    while !rem.is_empty() {
        let mut measure = Measure {
            content: string_buf.len()..=string_buf.len(),
            parent_line: parent_line_idx,
            index_on_parent_line: rlen(&measures),
        };
        loop {
            let rl_before = rem.len() as u32;
            let Ok((rem1, element)) = tab_element(rem, |to| {
                bend_targets.insert((line_in_part as u8, string_buf.len() as u32), to);
            }) else {
                break;
            };
            last_parsed_idx += rl_before - rem1.len() as u32; // multichar ticks
            rem = rem1;
            string_buf.push(RawTick { element });
            string_offsets_buf.push(start_source_offset + last_parsed_idx);
            if track_measures {
                measure.extend_1();
            }
        }
        if track_measures {
            measure.content = *measure.content.start()..=measure.content.end() - 1;
            string_measure_buf.push(measure);
            measures = *measures.start()..=measures.end() + 1;
        }
        rem = char('|')(rem)?.0;
        last_parsed_idx += 1;
    }
    // off by one: because we are using inclusive ranges, for example the first line, with only 1
    // measure, would be 0..=1 but we want it to be 0..=0
    if track_measures {
        measures = *measures.start()..=measures.end() - 1
    };
    Ok((
        rem,
        Partline {
            string_name,
            measures,
        },
    ))
}

/// A staff of a single string.
/// like `|--------------4-----------|`
/// The string it is on is encoded out-of-band
#[derive(Debug, PartialEq, Clone)]
pub struct Measure {
    /// The indices of the track this measure is on which belong to this measure
    pub content: RangeInclusive<usize>,
    pub parent_line: usize,
    pub index_on_parent_line: usize,
}

impl Measure {
    pub fn extend_1(&mut self) {
        self.content = *self.content.start()..=self.content.end() + 1
    }
    pub fn pop_1(&mut self) {
        self.content = *self.content.start()..=self.content.end() - 1
    }
    pub fn len(&self) -> usize {
        rlen(&self.content)
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
    pub fn get_content<'a>(&self, string_buf: &'a [RawTick]) -> &'a [RawTick] {
        &string_buf[self.content.clone()]
    }
    pub fn print_pretty_string(
        &self,
        string_buf: &[RawTick],
        track_idx: u8,
        bend_targets: &BendTargets,
    ) -> String {
        dump_ticks(&string_buf[self.content.clone()], track_idx, bend_targets)
    }
}
fn char(c: char) -> impl Fn(&str) -> Result<(&str, char), &str> {
    move |s: &str| match s.chars().next() {
        Some(cc) if cc == c => Ok((&s[1..], c)),
        _ => Err(s),
    }
}
pub fn dump_ticks(buf: &[RawTick], track_idx: u8, bend_offsets: &BendTargets) -> String {
    let mut pretty = String::new();
    for (tick_idx, x) in buf.iter().enumerate() {
        match x.element {
            TabElement::Fret(x) => pretty += &x.to_string(),
            TabElement::Rest => pretty += "-",
            TabElement::DeadNote => pretty += "x",
            TabElement::FretBend(x) => pretty += &format!("{x}b"),
            TabElement::FretBendTo(x) => {
                let y = bend_offsets
                    .get(&(track_idx, tick_idx as u32))
                    .expect("Unreachable: FretBendTo without target");
                pretty += &format!("{x}b{y}");
            }
        }
    }
    pretty
}

/// horribly inefficient, for debugging only
pub fn dump_tracks(tracks: &[Vec<RawTick>; 6], bend_targets: &BendTargets) -> String {
    let tick_cnt = tracks.iter().map(|x| x.len()).max().unwrap();
    let mut bufs = vec![String::new(); 6];
    for track in 0..6 {
        let mut i = 0;
        while i < tick_cnt {
            if i >= tracks[track].len() {
                break;
            };
            let tick_len = tracks
                .iter()
                .enumerate()
                .filter_map(|(track_idx, x)| x.get(i).map(|x| ((track_idx as u8, i as u32), x)))
                .map(|(pos, x)| x.element.repr_len(bend_targets, &pos))
                .max()
                .unwrap() as usize;
            use TabElement::*;
            match &tracks[track][i].element {
                Fret(x) => bufs[track].push_str(&format!("{x:<0$}", tick_len)),
                Rest => bufs[track].push_str(&format!("{1:<0$}", tick_len, "-")),
                DeadNote => bufs[track].push_str(&format!("{1:<0$}", tick_len, "x")),
                FretBend(x) => bufs[track].push_str(&format!("{x:<0$}b", tick_len - 1)),
                FretBendTo(x) => {
                    let y = bend_targets
                        .get(&(track as u8, i as u32))
                        .expect("Unreachable: FretBendTo without target");
                    bufs[track].push_str(&format!("{1:<0$}", tick_len, format!("{x}b{y}")))
                }
            };
            i += 1
        }
        bufs[track].push('\n');
    }
    bufs.concat()
}

fn string_name() -> impl Fn(&str) -> Result<(&str, char), &str> {
    move |s: &str| match s.chars().next() {
        Some(c) if c.is_alphabetic() => Ok((&s[1..], c)),
        _ => Err(s),
    }
}

#[derive(Debug, PartialEq, Clone)]
pub enum TabElement {
    Fret(u8),
    FretBend(u8),
    FretBendTo(u8),
    Rest,
    DeadNote,
}

#[inline(always)]
fn numeric(s: &str) -> Result<(&str, u8), &str> {
    let bytes = s.as_bytes();
    let mut i = 0;
    let mut sum = 0;
    while i < bytes.len() && bytes[i].is_ascii_digit() {
        sum *= 10;
        sum += bytes[i] - b'0';
        i += 1;
    }
    if i == 0 {
        return Err(s);
    };
    Ok((&s[i..], sum))
}

fn tab_element(s: &str, set_bend_target: impl FnOnce(u8)) -> Result<(&str, TabElement), &str> {
    let bytes = s.as_bytes();
    match bytes.first() {
        Some(b'-') => Ok((&s[1..], TabElement::Rest)),
        Some(b'x') => Ok((&s[1..], TabElement::DeadNote)),
        Some(48..=58) => {
            let (res, num) = numeric(s).unwrap();
            let bytes = res.as_bytes();
            if let Some(b'b') = bytes.first() {
                if let Ok((res, bend_target)) = numeric(&res[1..]) {
                    set_bend_target(bend_target);
                    return Ok((res, TabElement::FretBendTo(num)));
                }
                return Ok((&res[1..], TabElement::FretBend(num)));
            }
            Ok((res, TabElement::Fret(num)))
        }
        Some(_) | None => Err(s),
    }
}

impl TabElement {
    #[inline(always)]
    pub fn repr_len(&self, bend_targets: &BendTargets, pos: &(u8, u32)) -> u8 {
        match self {
            TabElement::Fret(x) => digit_cnt_u8(*x),
            TabElement::FretBend(x) => digit_cnt_u8(*x) + 1,
            TabElement::FretBendTo(x) => {
                let y = bend_targets
                    .get(pos)
                    .expect("TabElement::repr_len: FretBendTo without target");
                digit_cnt_u8(*x) + 1 + digit_cnt_u8(*y)
            }
            TabElement::Rest => 1,
            TabElement::DeadNote => 1,
        }
    }
}

#[derive(Debug, PartialEq, Clone)]
pub struct RawTick {
    pub element: TabElement,
}
