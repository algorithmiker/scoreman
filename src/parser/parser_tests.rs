use crate::parser::{parser2::parse2, partline};

#[test]
fn test_score() {
    let example_score = r#"
e|---|
B|-3-|
G|6-6|
D|---|
A|---|
E|---|

e|---|
B|3-3|
G|-6-|
D|---|
A|---|
E|---|
"#;
    insta::assert_debug_snapshot!(parse2(example_score.lines()));
}
#[test]
fn test_part() {
    let example_part = r#"
e|---|---|
B|-3-|3-3|
G|6-6|-6-|
D|---|---|
A|---|---|
E|---|---|"#;
    insta::assert_debug_snapshot!(parse2(example_part.lines()));
}

#[test]
fn test_partline() {
    insta::assert_debug_snapshot!(partline("e|--4-|-0--5-|", 0));
}
