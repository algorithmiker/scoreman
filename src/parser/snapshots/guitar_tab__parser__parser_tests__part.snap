---
source: src/parser/parser_tests.rs
expression: parse2(example_part.lines())
---
Ok(
    (
        [],
        Score(
            [
                Part {
                    part: [
                        Partline {
                            string_name: 'e',
                            measures: [
                                Measure {
                                    content: [
                                        RawTick {
                                            element: Rest,
                                            parent_line: 1,
                                            idx_on_parent_line: 2,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 1,
                                            idx_on_parent_line: 3,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 1,
                                            idx_on_parent_line: 4,
                                        },
                                    ],
                                    parent_line: 1,
                                    index_on_parent_line: 0,
                                },
                                Measure {
                                    content: [
                                        RawTick {
                                            element: Rest,
                                            parent_line: 1,
                                            idx_on_parent_line: 6,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 1,
                                            idx_on_parent_line: 7,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 1,
                                            idx_on_parent_line: 8,
                                        },
                                    ],
                                    parent_line: 1,
                                    index_on_parent_line: 0,
                                },
                            ],
                        },
                        Partline {
                            string_name: 'B',
                            measures: [
                                Measure {
                                    content: [
                                        RawTick {
                                            element: Rest,
                                            parent_line: 2,
                                            idx_on_parent_line: 2,
                                        },
                                        RawTick {
                                            element: Fret(
                                                3,
                                            ),
                                            parent_line: 2,
                                            idx_on_parent_line: 3,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 2,
                                            idx_on_parent_line: 4,
                                        },
                                    ],
                                    parent_line: 2,
                                    index_on_parent_line: 0,
                                },
                                Measure {
                                    content: [
                                        RawTick {
                                            element: Fret(
                                                3,
                                            ),
                                            parent_line: 2,
                                            idx_on_parent_line: 6,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 2,
                                            idx_on_parent_line: 7,
                                        },
                                        RawTick {
                                            element: Fret(
                                                3,
                                            ),
                                            parent_line: 2,
                                            idx_on_parent_line: 8,
                                        },
                                    ],
                                    parent_line: 2,
                                    index_on_parent_line: 0,
                                },
                            ],
                        },
                        Partline {
                            string_name: 'G',
                            measures: [
                                Measure {
                                    content: [
                                        RawTick {
                                            element: Fret(
                                                6,
                                            ),
                                            parent_line: 3,
                                            idx_on_parent_line: 2,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 3,
                                            idx_on_parent_line: 3,
                                        },
                                        RawTick {
                                            element: Fret(
                                                6,
                                            ),
                                            parent_line: 3,
                                            idx_on_parent_line: 4,
                                        },
                                    ],
                                    parent_line: 3,
                                    index_on_parent_line: 0,
                                },
                                Measure {
                                    content: [
                                        RawTick {
                                            element: Rest,
                                            parent_line: 3,
                                            idx_on_parent_line: 6,
                                        },
                                        RawTick {
                                            element: Fret(
                                                6,
                                            ),
                                            parent_line: 3,
                                            idx_on_parent_line: 7,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 3,
                                            idx_on_parent_line: 8,
                                        },
                                    ],
                                    parent_line: 3,
                                    index_on_parent_line: 0,
                                },
                            ],
                        },
                        Partline {
                            string_name: 'D',
                            measures: [
                                Measure {
                                    content: [
                                        RawTick {
                                            element: Rest,
                                            parent_line: 4,
                                            idx_on_parent_line: 2,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 4,
                                            idx_on_parent_line: 3,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 4,
                                            idx_on_parent_line: 4,
                                        },
                                    ],
                                    parent_line: 4,
                                    index_on_parent_line: 0,
                                },
                                Measure {
                                    content: [
                                        RawTick {
                                            element: Rest,
                                            parent_line: 4,
                                            idx_on_parent_line: 6,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 4,
                                            idx_on_parent_line: 7,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 4,
                                            idx_on_parent_line: 8,
                                        },
                                    ],
                                    parent_line: 4,
                                    index_on_parent_line: 0,
                                },
                            ],
                        },
                        Partline {
                            string_name: 'A',
                            measures: [
                                Measure {
                                    content: [
                                        RawTick {
                                            element: Rest,
                                            parent_line: 5,
                                            idx_on_parent_line: 2,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 5,
                                            idx_on_parent_line: 3,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 5,
                                            idx_on_parent_line: 4,
                                        },
                                    ],
                                    parent_line: 5,
                                    index_on_parent_line: 0,
                                },
                                Measure {
                                    content: [
                                        RawTick {
                                            element: Rest,
                                            parent_line: 5,
                                            idx_on_parent_line: 6,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 5,
                                            idx_on_parent_line: 7,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 5,
                                            idx_on_parent_line: 8,
                                        },
                                    ],
                                    parent_line: 5,
                                    index_on_parent_line: 0,
                                },
                            ],
                        },
                        Partline {
                            string_name: 'E',
                            measures: [
                                Measure {
                                    content: [
                                        RawTick {
                                            element: Rest,
                                            parent_line: 6,
                                            idx_on_parent_line: 2,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 6,
                                            idx_on_parent_line: 3,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 6,
                                            idx_on_parent_line: 4,
                                        },
                                    ],
                                    parent_line: 6,
                                    index_on_parent_line: 0,
                                },
                                Measure {
                                    content: [
                                        RawTick {
                                            element: Rest,
                                            parent_line: 6,
                                            idx_on_parent_line: 6,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 6,
                                            idx_on_parent_line: 7,
                                        },
                                        RawTick {
                                            element: Rest,
                                            parent_line: 6,
                                            idx_on_parent_line: 8,
                                        },
                                    ],
                                    parent_line: 6,
                                    index_on_parent_line: 0,
                                },
                            ],
                        },
                    ],
                    begin_line_idx: 1,
                    end_line_idx: 6,
                },
            ],
        ),
    ),
)
